core/trino-main/src/main/java/io/trino/execution/scheduler/SqlQueryScheduler.java

```diff
@@ -1042,6 +1040,9 @@ public void schedule()
                         ImmutableMultimap.of());
                 stageExecution.schedulingComplete();
                 remoteTask.ifPresent(task -> coordinatorTaskManager.addSourceTaskFailureListener(task.getTaskId(), failureReporter));
+                if (queryStateMachine.getQueryState() == STARTING && remoteTask.isPresent()) {
```

Comment:
why just `STARTING`?

Reply:
The order of states is `STARTING` -> `RUNNING` -> `FINSHING -> FINISHED | FAILED`. If it's already in a later stage there's no need to transition

Reply:
Technically you do not need `if (queryStateMachine.getQueryState() == STARTING)` at all as `queryStateMachine.transitionToRunning` does the check (it does `queryState.setIf(RUNNING, currentState -> currentState.ordinal() < RUNNING.ordinal());`)